fix(cpp): incorrect indent of class with opening { on new line

jedrzejboczar opened the pull request:
Removed `@indent.begin` from `(class_specifier)`, otherwise it indents as:
```cpp
class Foo
    {
        int x;
        class Bar {
            int y;
        };
        Bar z;
    };
```
Added a test for this.

I also removed  `@indent.begin`  from `(condition_clause)` as I couldn't find why this is there. No tests fail due to removing it and I tried some variants of `if (x) ...` and indenting still seems to work correctly, so maybe this was some legacy code that is no longer needed due to changes in `c/indents.scm`?

lucario387 commented:
When is `condition_clause` appearing though? I do not know about C++. Last time I fixed `c/indents.scm`, I just made sure that all C++ tests also passed, I did not check what is needed/not needed in C++

lucario387 commented on tests/indent/cpp/class_newline.cpp:
:+1:

jedrzejboczar commented:
Here are all the `condition_clause` occurrences: https://github.com/tree-sitter/tree-sitter-cpp/blob/master/grammar.js#L770-L788, so it's all just valid C.

All the following cases seem to still work for me (`switch` without braces doesn't work but I think it's invalid syntax):
```cpp
if (1)
    foo();

if (1) {
    foo();
}

if (1)
{
    foo();
}

switch (1) {
    case 1: break;
    default: break;
}

switch (1) 
{
    case 1: break;
    default: break;
}

while (1) 
    foo();


while (1) {
    foo();
}

while (1)
{
    foo();
}
```

lucario387 commented:
We'll need to keep `condition_clause`, or adding a pattern to support mutliline switch condition in C (if any more than 1% of C programmer do that I'd be surprise)

If you deem that it's not worth it to keep it then it's ok to remove

jedrzejboczar commented:
This is a syntax error in treesitter parser (and possibly for a compiler too?) ![Screenshot from 2023-12-20 12-09-36](https://github.com/nvim-treesitter/nvim-treesitter/assets/16623787/a8378ec3-f65b-4a49-b3cb-3ca97b73497a)

jedrzejboczar commented:
Ok, suprisingly this is valid C... `gcc -Wall test.cpp -o test` succeeds. Still I think I've never seen something like this and it makes no sense to write a switch with a single case, so I wouldn't care.

lucario387 commented:
> This is a syntax error in treesitter parser (and possibly for a compiler too?)
> 

I mean the `switch (condition_clause)`. It's obvious not having {} is not allowed

lucario387 commented:
I'd vote for keeping condition_clause, while class_specifier needs to go.

jedrzejboczar commented:
But `condition_clause` is the node with condition that is in every switch/if/while, e.g. here it is `(1)`
![image](https://github.com/nvim-treesitter/nvim-treesitter/assets/16623787/e0bfb356-7db0-41a5-8eaa-3d8221c72030)
So even if we should keep `condition_clause` then I think it should be in `c/indents.scm` not in cpp, as this is just C syntax.

jedrzejboczar commented:
But it looks to me like it is not needed, something else is indenting all those if/while/switch statements.

lucario387 commented:
It's for indenting inside the condtions. For if/else/while I did it from C. For switch's conditions there's nothing yet

jedrzejboczar commented:
Ok, I reverted removing of the `condition_clause` as I'm not sure where I should add it in c/indents.scm.

But overall, I think it is not used anyway. I added print statements in `indent.lua` to see which nodes are used to increase the indent and then tried indenting line 4 (using `=` operator) in the following code - it looks like it just uses `compound_statement` (so the curly braces `{ ... }`):
![Screenshot from 2023-12-21 20-28-11](https://github.com/nvim-treesitter/nvim-treesitter/assets/16623787/00dc2519-9abc-43fd-8ab9-26fe2e80c4f8)

lucario387 commented:
It's very niche, but I like this version.
